### ISSUE TITLE
Add the ability to monitor and display amount of filament remaining on a spool

### ISSUE THREAD

**Issue by julie777** [bambulab/BambuStudio#2386]
## Is your feature request related to a problem? Please describe.

Sometimes I do not have enough filament remaining to print a part, and I do not have another spool of that filament available to be able to swap spools mid print. I currently have to make a guess whether to try the print and hope for the best, or just wait until I buy more filament.

Another use case is when printing overnight or remotely and I do not want to start a print that cannot finish.

Note: I do not have an AMS, and even if I did the AMS does not track non-BBL filament without RFID tags.

## Describe the solution you'd like

In Bambu Studio provide the ability to number filament spools and have it use the estimated amount of filament for a model to track usage for that spool.

The implementation of this is simple and does not require being extremely accurate to be useful.

### Supporting Structure

Add a numbered spool list to the Bambu Studio configuration to keep track of the known spools and current weight.

### User Interface
1. In the print dialog provide spool number drop-down menu to allow choosing which spool is currently being used. The default value should be the last spool used. (It is common to print multiple jobs one after another using the same spool.)
2. The last option in the menu should be add new spool which provides a way to enter the starting weight.
3. In the chooser, the current calculated weight should be shown next to the spool number. 
4. if the weight for the current spool is less than the amount needed to print then the user should be warned before printing. However, the user should be able to continue if desired. The printer will run out, stop and notify the user. The user will load a new spool. When printing continues, Bambu Studio should prompt for the spool number of the inserted spool and continue based on that spool.
5. allow the user to remove a spool number from the list, or reset the weight (if reusing the same number is allowed)

### Notes
- because it might be difficult to know how much filament has been use up to the point where the filament runs out and is changed by the user, that part could be optional. However, given that the printer is reporting the current layer number, I believe that Bambu Studio can calculate this.
- better accuracy can be achieved if the estimated amount of filament for a print job includes any required filament during the startup code. I do not know if that is in the estimated amount of filament currently.
- Optionally, provide the ability to set the current spool number for a given printer in the "prepare" section so I can just set it after I change spools and not have to wait until I am sending the next print.

## Describe alternatives you've considered

I could weight my spools full and empty and keep my own list and update it manually.

**Comment by SnickCH** [comment]
+1
It shouldn't be to hard to calculate how much it is left. Maybe it would be easier with the length because then it is not important about the density of the filament
- initial weight or length (set by the user)
- bambustudio estimates the weight / length of the print (+ some margin) and removes it from the current weight / lenght

**Comment by ShawnLeary** [comment]
Would love if AMS had a general tracker setting.  If new spool it should have an idea of how much is left.

**Comment by EA0003** [comment]
Yes! this would be great, specially since the weight is really what matters. Even if they add a "safety" factor of 5% which the user can adjust, that'd be so nice

**Comment by theukedge** [comment]
+1. Especially with AMS, Bambu Studio ought to be able to track how much filament is left on each spool and allow the user to view that information and/or provide a warning if you're trying to print something that you don't have enough filament for.

**Comment by cfuller007** [comment]
I would love this. I would love to track how much each print cost me in price and filament. Additionally, this would allow me to be ready when I need to order more filament from bambú and how much I have left.

**Comment by Maitresinh** [comment]
+1

**Comment by PnsDev** [comment]
huge +1

**Comment by cameronnewell** [comment]
+1

**Comment by prioa** [comment]
+1

**Comment by MathiasSven** [comment]
+1! But also, isn't this somewhat a duplicate of #1186

**Comment by mike-masson** [comment]
> +1! But also, isn't this somewhat a duplicate of #1186

+1 

Yes, but that seems like an idea in which they create an entire indexing database, which would be awesome and the CEO did mention something like this in the AMA that he did over on reddit. Whereas this seems more like a quick addition as it should be a relatively easy thing to do since Bambu Studio already estimates remaining filament by measuring how much filament is pulled per rotation of the spool. I'm not sure what the margin of error is, but even 10% would be useful.

**Comment by jjvelar** [comment]
Hi,
This will be a great adding for AMS Lite.
Is this in the pipeline?

**Comment by wawibu** [comment]
+1 
I'm new to Bambu and would like to see the remaining length. Others - like Octoprint - could provide that information. No need to be 100% accurate, I would love to get an indication on how much is left so that I know before I print if that spool would have enough or not

**Comment by ydinopoulos** [comment]
+1

**Comment by Csaba94** [comment]
+1

**Comment by Anthony-Bec** [comment]
+1

**Comment by composerken** [comment]
+1

**Comment by FaberO42** [comment]
Big +1 for something along these lines.
Bambu great for managing 3D printing process, but significant room for improvement in managing key (filament) consumables use.
Also requires to support more than just Bambu brand filaments.

**Comment by MiragonMx** [comment]
+1

**Comment by buchpat** [comment]
+1

**Comment by cyberporkexception** [comment]
+1

**Comment by armas01** [comment]
+1

**Comment by fclarat** [comment]
+1

**Comment by chunkahunk** [comment]
+1 please

**Comment by blummarci24** [comment]
+1

**Comment by NiklasRichter2222** [comment]
+1

**Comment by daveisfera** [comment]
I believe this is a duplicate of #1887

**Comment by blahnika** [comment]
+2

**Comment by jhenley01** [comment]
As someone who runs a small print farm this would be great to have. 

I typically load a spool and then print until it's empty so I'd be happy with just a simple counter that adds up the amount of filament that's been used since the filament was loaded. Right now I have to keep a dry erase board by each printer and subtract the amount every time i start a print, would be great if this was built into the printer somewhere.

**Comment by r-nd-m** [comment]
+1
